wrappers/binaries.go

```diff
@@ -21,13 +21,71 @@
 package wrappers
 
 import (
+	"bufio"
 	"os"
+	"regexp"
+	"strconv"
 
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/snap"
 )
 
+// detectLegacy checks whether bash-completion is version 2.2 or later.
+// Prior to 2.2, bash-completion does not support XDG_DATA_DIRS.
+// If it fails to detect the version of bash-completion, it disables
+// completion.
+func detectLegacy() (noCompletion bool, legacyCompletion bool) {
+	fd, err := os.Open(dirs.BashCompletionScript)
+	if err != nil {
+		// Cannot read file, disable completion
+		return true, false
+	}
+	defer fd.Close()
+
+	releaseOld := regexp.MustCompile(`^# *RELEASE: ([0-9]+)[.]([0-9]+)$`)
+	releaseNew := regexp.MustCompile(`^BASH_COMPLETION_VERSINFO=`)
+	s := bufio.NewScanner(fd)
+	var matched []string
+	for s.Scan() {
+		line := s.Text()
+		matched = releaseOld.FindStringSubmatch(line)
+		if matched == nil {
+			if releaseNew.MatchString(line) {
+				// It must be 2.6 or later
+				return false, false
+			}
+		} else {
+			break
+		}
+	}
+	if err := s.Err(); err != nil {
+		// Cannot read file, disable completion
+		return true, false
+	}
+	if matched == nil {
+		// Unknown version: disable completion
+		return true, false
+	}
+
+	major, majorErr := strconv.Atoi(matched[1])
+	minor, minorErr := strconv.Atoi(matched[2])
+	if majorErr != nil || minorErr != nil {
+		// Cannot parse version, disable completion
+		return true, false
+	}
+
+	if major < 2 {
+		return false, true
+	}
+
+	if major > 2 {
+		return false, false
+	}
+
+	return false, minor < 2
```

Comment:
In practice, because 3.1 will not reach this part of the code, that would work. Although not in theory. But this is not how we compare versions. Any time someone would read that part of the code would see `major >= 2 && minor >= 2`, they would think there is a bug, until they figure out that major cannot be more than 2, in that part of the code... in practice, but not in theory.

Reply:
Well, yes code is structured in a way that for 2.6+ (or at least any version that has BASH_COMPLETION_VERSINFO=) we just return early and the rest isn't even reached, so the check `if major > 2` is actually impossible to reach, isn't it?
Perhaps there should be a comment after the for loop indicating that  the rest is reached only if the bash-completion had the old string and now we need to check the version which can be up to 2.5 really. FWIW if any future version drops the line we match on, the code will be incorrect anyway.

Reply:
I will try to rewrite the code a different way. @mardy also had comments on this function anyway. But I cannot write `major >= 2 && minor >= 2`, that burns my eyes.

Reply:
I have changed to code to use `strutil.CompareVersion`.